File: fast-chat/go/internal/chat/message.go
package chat

import (
	"encoding/json"
	"html"
	"strconv"
	"time"
)

// Message represents a single chat message
type Message struct {
	ID        int64     `json:"id"`
	Nickname  string    `json:"nickname"`
	Text      string    `json:"text"`
	Timestamp time.Time `json:"timestamp"`
}

// NewMessage creates a new message with HTML-escaped content
func NewMessage(id int64, nickname, text string) *Message {
	return &Message{
		ID:        id,
		Nickname:  html.EscapeString(nickname),
		Text:      html.EscapeString(text),
		Timestamp: time.Now(),
	}
}

// ClientMessage is the JSON wire format sent to clients
type ClientMessage struct {
	Type      string      `json:"type"`   // "message", "system", "usercount", "history"
	Action    string      `json:"action"` // "append", "prepend", "replace"
	Data      interface{} `json:"data"`
	Timestamp int64       `json:"timestamp"` // Unix milliseconds for client latency measurement
}

type MessageData struct {
	ID       string `json:"id"`
	Nickname string `json:"nickname"`
	Text     string `json:"text"`
	Time     string `json:"time"`
	IsSystem bool   `json:"isSystem,omitempty"`
	ClientID string `json:"clientId,omitempty"`
	ServerID string `json:"serverId,omitempty"`
}

// ToJSON converts the message to JSON format
func (m *Message) ToJSON() string {
	msg := ClientMessage{
		Type:      "message",
		Action:    "append",
		Timestamp: time.Now().UnixMilli(),
		Data: MessageData{
			ID:       strconv.FormatInt(m.ID, 10),
			Nickname: m.Nickname,
			Text:     m.Text,
			Time:     m.Timestamp.Format("15:04:05.000"),
		},
	}

	bytes, err := json.Marshal(msg)
	if err != nil {
		return `{"type":"error","data":"Failed to encode message"}`
	}
	return string(bytes)
}

// ToHTML renders the message as an HTML fragment (kept for compatibility)
func (m *Message) ToHTML() string {
	return `<div data-target="#messages" data-action="append">
  <div class="msg" data-id="` + formatInt64(m.ID) + `">
    <span class="nick">` + m.Nickname + `</span>
    <span class="text">` + m.Text + `</span>
    <time>` + m.Timestamp.Format("15:04:05.000") + `</time>
  </div>
</div>`
}

// SystemMessageJSON creates a system message in JSON format
func SystemMessageJSON(text string) string {
	msg := ClientMessage{
		Type:      "message",
		Action:    "append",
		Timestamp: time.Now().UnixMilli(),
		Data: MessageData{
			Text:     html.EscapeString(text),
			IsSystem: true,
		},
	}

	bytes, err := json.Marshal(msg)
	if err != nil {
		return `{"type":"error","data":"Failed to encode system message"}`
	}
	return string(bytes)
}

// SystemMessage creates a system message HTML fragment (kept for compatibility)
func SystemMessage(text string) string {
	return `<div data-target="#messages" data-action="append">
  <div class="sys">` + html.EscapeString(text) + `</div>
</div>`
}

// UserCountJSON creates a user count update in JSON format
func UserCountJSON(count int) string {
	plural := ""
	if count != 1 {
		plural = "s"
	}

	msg := ClientMessage{
		Type:      "usercount",
		Action:    "replace",
		Timestamp: time.Now().UnixMilli(),
		Data:      strconv.Itoa(count) + " user" + plural + " online",
	}

	bytes, err := json.Marshal(msg)
	if err != nil {
		return `{"type":"error","data":"Failed to encode user count"}`
	}
	return string(bytes)
}

// UserCountHTML creates a user count update HTML fragment (kept for compatibility)
func UserCountHTML(count int) string {
	plural := ""
	if count != 1 {
		plural = "s"
	}
	return `<span data-target="#user-count" data-action="replace">` +
		formatInt(count) + ` user` + plural + ` online</span>`
}

// Helper to format int64 to string without imports
func formatInt64(n int64) string {
	if n == 0 {
		return "0"
	}

	negative := n < 0
	if negative {
		n = -n
	}

	var buf [20]byte
	i := len(buf) - 1

	for n > 0 {
		buf[i] = byte('0' + n%10)
		n /= 10
		i--
	}

	if negative {
		buf[i] = '-'
		i--
	}

	return string(buf[i+1:])
}

// Helper to format int to string
func formatInt(n int) string {
	return formatInt64(int64(n))
}

File: fast-chat/rust/src/chat/message.rs
use chrono::{DateTime, Utc};
use serde::{Deserialize, Serialize};

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct Message {
    pub id: i64,
    pub nickname: String,
    pub text: String,
    pub timestamp: DateTime<Utc>,
}

impl Message {
    pub fn new(id: i64, nickname: String, text: String) -> Self {
        Self {
            id,
            nickname: html_escape::encode_text(&nickname).to_string(),
            text: html_escape::encode_text(&text).to_string(),
            timestamp: Utc::now(),
        }
    }

    pub fn to_json(&self) -> String {
        let msg = ClientMessage {
            r#type: "message".to_string(),
            action: "append".to_string(),
            timestamp: Utc::now().timestamp_millis(),
            data: serde_json::json!(MessageData {
                id: self.id.to_string(),
                nickname: self.nickname.clone(),
                text: self.text.clone(),
                time: self.timestamp.format("%H:%M:%S%.3f").to_string(),
                is_system: None,
                client_id: None,
                server_id: None,
            }),
        };

        serde_json::to_string(&msg).unwrap_or_else(|_| {
            r#"{"type":"error","data":"Failed to encode message"}"#.to_string()
        })
    }
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct ClientMessage {
    pub r#type: String,
    pub action: String,
    pub data: serde_json::Value,
    pub timestamp: i64,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct MessageData {
    pub id: String,
    pub nickname: String,
    pub text: String,
    pub time: String,
    #[serde(skip_serializing_if = "Option::is_none")]
    #[serde(rename = "isSystem")]
    pub is_system: Option<bool>,
    #[serde(skip_serializing_if = "Option::is_none")]
    #[serde(rename = "clientId")]
    pub client_id: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    #[serde(rename = "serverId")]
    pub server_id: Option<String>,
}

pub fn create_echo_message(msg: &Message, client_id: String) -> String {
    let echo_msg = ClientMessage {
        r#type: "message".to_string(),
        action: "append".to_string(),
        timestamp: Utc::now().timestamp_millis(),
        data: serde_json::json!(MessageData {
            id: msg.id.to_string(),
            nickname: msg.nickname.clone(),
            text: msg.text.clone(),
            time: msg.timestamp.format("%H:%M:%S%.3f").to_string(),
            is_system: None,
            client_id: Some(client_id),
            server_id: Some(msg.id.to_string()),
        }),
    };

    serde_json::to_string(&echo_msg).unwrap_or_else(|_| {
        r#"{"type":"error","data":"Failed to encode echo message"}"#.to_string()
    })
}

pub fn system_message_json(text: &str) -> String {
    let msg = ClientMessage {
        r#type: "message".to_string(),
        action: "append".to_string(),
        timestamp: Utc::now().timestamp_millis(),
        data: serde_json::json!(MessageData {
            id: String::new(),
            nickname: String::new(),
            text: html_escape::encode_text(text).to_string(),
            time: String::new(),
            is_system: Some(true),
            client_id: None,
            server_id: None,
        }),
    };

    serde_json::to_string(&msg).unwrap_or_else(|_| {
        r#"{"type":"error","data":"Failed to encode system message"}"#.to_string()
    })
}

pub fn user_count_json(count: usize) -> String {
    let plural = if count != 1 { "s" } else { "" };
    let text = format!("{} user{} online", count, plural);

    let msg = ClientMessage {
        r#type: "usercount".to_string(),
        action: "replace".to_string(),
        timestamp: Utc::now().timestamp_millis(),
        data: serde_json::json!(text),
    };

    serde_json::to_string(&msg).unwrap_or_else(|_| {
        r#"{"type":"error","data":"Failed to encode user count"}"#.to_string()
    })
}

pub fn history_messages_json(messages: Vec<&Message>) -> String {
    let history_data: Vec<MessageData> = messages
        .iter()
        .map(|msg| MessageData {
            id: msg.id.to_string(),
            nickname: msg.nickname.clone(),
            text: msg.text.clone(),
            time: msg.timestamp.format("%H:%M:%S%.3f").to_string(),
            is_system: None,
            client_id: None,
            server_id: None,
        })
        .collect();

    let msg = ClientMessage {
        r#type: "history".to_string(),
        action: "prepend".to_string(),
        timestamp: Utc::now().timestamp_millis(),
        data: serde_json::json!(history_data),
    };

    serde_json::to_string(&msg).unwrap_or_else(|_| {
        r#"{"type":"error","data":"Failed to encode history"}"#.to_string()
    })
}
